[36243][30226] Different avatars for placeholders and groups

oliverguenther opened the pull request:
This PR:

- [x] allows rendering different avatar styles based on the type of a principal.
- [x] Extends the CreateAutocompleter as an intermediate solution as discussed to allow avatar rendering

The styles are taking as is from Wieland so I'd rather not change them in the course of this PR.

https://community.openproject.org/wp/36243
https://community.openproject.org/wp/30226/

oliverguenther committed:
Fix members autocompleter to allow hrefs

oliverguenther committed:
Add rendering of different avatar types

oliverguenther committed:
Remove unused code in backend avatar rendering

oliverguenther committed:
Add very basic test that the avatar gets customized per type

oliverguenther commented:
Hi @b12f , could you take a look at the frontend changes here? In the backend, the avatar helper patch you've also worked on was drastically simplified as we always render the op-principal component anyway, so we don't have to check what avatar type the user actually has.

I also ensured that the members autocompleter returns the correct hrefs we expected in the frontend so there's some minor changes there but that's otherwise only a move of where the json gets rendered in the backend.

oliverguenther committed:
Allow avatar rendering in the CreateAutocompleter

b12f commented on frontend/src/global_styles/content/modules/_avatars.sass:
```diff
@@ -28,4 +28,12 @@
\ No newline at end of file
+  margin-bottom: 2px
+
+.avatar
```
if this is a BEM block you want to keep I'd suggest renaming it to `op-avatar` and putting it into its own file

b12f replied:
Might be hard to do right now with all the legacy stuff though

b12f commented on frontend/src/global_styles/content/modules/_avatars.sass:
```diff
@@ -28,4 +28,12 @@
+  margin-bottom: 2px
+
+.avatar
+  &_placeholder_user
```
```suggestion
  &_placeholder-user
```

An underscore denotes a modifier, of which you can only have one per class

oliverguenther replied:
I was simply reusing the existing principal types that use underscores. Not too hard to just replace() them though. Will do

oliverguenther committed:
Restructure avatars into BEM css

oliverguenther committed:
Remove avatar-classes in favor of passing size from backend

oliverguenther committed:
Remove  to avoid confusion

We use tracking with both href and , and  was introduced

to avoid a property called href to be overridden, but that was never a problem before.

So from now we can safely use .href

oliverguenther committed:
Fix broken principal avatar in user activity

oliverguenther commented:
@b12f I extracted all user avatar related styles into two BEM blocks: op-principal and op-principal-avatar and addressed your feedback. 

I removed the ` avatar-classes` input to principals as that was almost always used for setting the classes. The backend now simply uses the `size` property instead, so we can safely remove it. There were two cases that used the class to override some positioning. For them I allowed a limited override to `.op-principal-avatar` to avoid too much restyling in this PR.

I also noticed we were confusing `$href` and `href` in the value options of CreateAutocompleter and the Principal inputs. So I removed the accessor to $href  for all hal resources as it's no longer necessary.

oliverguenther committed:
Ensure filters get updated even if the form is not reloaded

There was a bug in conditionallyLoadForm that checked the  property of link (always undefined)

and would thus always reload a form.

What we actually need is a form that gets reloaded only when the query link changes, but still update

services from the updated query, such as filters

oliverguenther committed:
Adapt specs to changes classes

oliverguenther committed:
Rename op-principal-avatar to op-avatar

b12f reviewed (approved):
LGTM
